ProjFS.Mac/PrjFSKext/VnodeCache.cpp

```diff
@@ -0,0 +1,381 @@
+                
+                    UpgradeToExclusiveLock(s_entriesLock);
+                    lockElevatedToExclusive = true;
+                    
+                    UpdateCacheEntryToLatest_ExclusiveLocked(
+                        perfTracer,
+                        cacheMissFallbackFunctionCounter,
+                        cacheMissFallbackFunctionInnerLoopCounter,
+                        cacheIndex,
+                        vnode,
+                        vnodeFsidInode,
+                        vnodeVid);
+                }
+                else
+                {
+                    perfTracer->IncrementCount(cacheHitCounter, true /*ignoreSampling*/);
+                }
+                
+                rootHandle = s_entries[cacheIndex].virtualizationRoot;
+            }
+            else
+            {
+                perfTracer->IncrementCount(cacheMissCounter, true /*ignoreSampling*/);
+            
+                UpgradeToExclusiveLock(s_entriesLock);
+                lockElevatedToExclusive = true;
+            
+                if (!FindAndUpdateEntryToLatest_ExclusiveLocked(
+                        perfTracer,
+                        cacheMissFallbackFunctionCounter,
+                        cacheMissFallbackFunctionInnerLoopCounter,
+                        vnode,
+                        vnodeFsidInode,
+                        cacheIndex,    // starting index
+                        indexFromHash, // stopping index
+                        vnodeVid,
+                        invalidateEntry,
+                        /* out */ rootHandle))
+                {
+                    ClearCacheAndInsertEntry_ExclusiveLocked(
+                        perfTracer,
+                        cacheMissFallbackFunctionCounter,
+                        cacheMissFallbackFunctionInnerLoopCounter,
+                        vnode,
+                        vnodeFsidInode,
+                        indexFromHash,
+                        vnodeVid,
+                        /* out */ rootHandle);
+                }
+            }
+        }
+        else
+        {
+            perfTracer->IncrementCount(cacheMissCounter, true /*ignoreSampling*/);
+            
+            UpgradeToExclusiveLock(s_entriesLock);
+            lockElevatedToExclusive = true;
+            
+            ClearCacheAndInsertEntry_ExclusiveLocked(
+                perfTracer,
+                cacheMissFallbackFunctionCounter,
+                cacheMissFallbackFunctionInnerLoopCounter,
+                vnode,
+                vnodeFsidInode,
+                indexFromHash,
+                vnodeVid,
+                /* out */ rootHandle);
+        }
+    }
+    
+    if (lockElevatedToExclusive)
+    {
+        RWLock_ReleaseExclusive(s_entriesLock);
+    }
+    else
+    {
+        RWLock_ReleaseShared(s_entriesLock);
+    }
+    
+    return rootHandle;
+}
+
+void VnodeCache_InvalidateCache(PerfTracer* _Nullable perfTracer)
+{
+    if (perfTracer)
```

Comment:
Style nit: so far we've been avoiding implicit pointer-to-bool conversions, opting instead to compare explicitly with `nullptr`.